summingbird-storm/src/main/scala/com/twitter/summingbird/storm/spout/KeyValueSpout.scala

```diff
@@ -4,32 +4,63 @@ import backtype.storm.spout.SpoutOutputCollector
 import backtype.storm.task.TopologyContext
 import backtype.storm.topology.{ IRichSpout, OutputFieldsDeclarer }
 import backtype.storm.tuple.Fields
+import com.twitter.algebird.Semigroup
+import com.twitter.summingbird.online.Externalizer
+import com.twitter.summingbird.online.executor.KeyValueShards
+import com.twitter.summingbird.online.option.SummerBuilder
 import com.twitter.summingbird.storm.Constants._
 import com.twitter.tormenta.spout.SpoutProxy
 import java.util
 import java.util.{ List => JList }
-import com.twitter.summingbird.storm.collector.TransformingOutputCollector
+import scala.collection.mutable.{ MutableList => MList }
+import com.twitter.summingbird.storm.collector.{ AggregatorOutputCollector, TransformingOutputCollector }
 
 /**
  * This is a spout used when the spout is being followed by summer.
- * It uses a TransformingOutputCollector on open.
+ * It uses a AggregatorOutputCollector on open.
  */
 
-class KeyValueSpout(in: IRichSpout) extends SpoutProxy {
+class KeyValueSpout[K, V: Semigroup](val in: IRichSpout, summerBuilder: SummerBuilder, summerShards: KeyValueShards, @transient callOnOpen: (TopologyContext) => Unit) extends SpoutProxy {
+
+  private var adapterCollector: AggregatorOutputCollector[K, V] = _
+  val lockedFn = Externalizer(callOnOpen)
+  var lastDump = System.currentTimeMillis()
 
   override def declareOutputFields(declarer: OutputFieldsDeclarer) = {
     declarer.declare(new Fields(AGG_KEY, AGG_VALUE))
   }
 
-  /*
-  * The transform is the function which unwraps the Value object to get the actual fields present in it.
-  */
-
   override def open(conf: util.Map[_, _],
     topologyContext: TopologyContext,
     outputCollector: SpoutOutputCollector): Unit = {
-    val adapterCollector = new TransformingOutputCollector(outputCollector, _.get(0).asInstanceOf[JList[AnyRef]])
-    self.open(conf, topologyContext, adapterCollector)
+    adapterCollector = new AggregatorOutputCollector(outputCollector, _.get(0).asInstanceOf[JList[AnyRef]], summerBuilder, summerShards)
+    lockedFn.get(topologyContext)
+    in.open(conf, topologyContext, adapterCollector)
+  }
+
+  override def nextTuple(): Unit = {
+    if (System.currentTimeMillis() - lastDump > 1000) {
+      adapterCollector.timerFlush()
+      lastDump = System.currentTimeMillis()
+    }
+    in.nextTuple()
+  }
+
+  override def ack(msgId: Object): Unit = {
```

Comment:
Could you comment who calls ack with a msgId which is actually a list of message ids. Seems pretty unintuitive so I want to understand for my own knowledge.